From 629bd94e07c09b1800752d23bf5943fa2244d0e3 Mon Sep 17 00:00:00 2001
From: Luca Fancellu <luca.fancellu@arm.com>
Date: Thu, 15 Jun 2023 16:44:35 +0100
Subject: [PATCH] net: eth: Fix ethernet max header size on VLAN presence

Currently NET_ETH_MAX_HDR_SIZE is computed as the size of the
structure struct net_eth_hdr, but this does not take into account
the additional 4 bytes on the ethernet frame header in the presence
of VLAN, fix the issue using the size of struct net_eth_vlan_hdr
when CONFIG_NET_VLAN is enabled.

Signed-off-by: Luca Fancellu <luca.fancellu@arm.com>
Upstream-Status: Backport [main: b3f5db0c718f76560238a4e4c06bb5d8619589b9]
---
 include/zephyr/net/ethernet.h | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/include/zephyr/net/ethernet.h b/include/zephyr/net/ethernet.h
index 0807c2be33..b4fad4224d 100644
--- a/include/zephyr/net/ethernet.h
+++ b/include/zephyr/net/ethernet.h
@@ -96,8 +96,14 @@ struct net_eth_addr {
 
 #define NET_ETH_MINIMAL_FRAME_SIZE	60
 #define NET_ETH_MTU			1500
-#define _NET_ETH_MAX_FRAME_SIZE	(NET_ETH_MTU + sizeof(struct net_eth_hdr))
+
+#if defined(CONFIG_NET_VLAN)
+#define _NET_ETH_MAX_HDR_SIZE		(sizeof(struct net_eth_vlan_hdr))
+#else
 #define _NET_ETH_MAX_HDR_SIZE		(sizeof(struct net_eth_hdr))
+#endif
+
+#define _NET_ETH_MAX_FRAME_SIZE	(NET_ETH_MTU + _NET_ETH_MAX_HDR_SIZE)
 /*
  * Extend the max frame size for DSA (KSZ8794) by one byte (to 1519) to
  * store tail tag.
-- 
2.34.1

